h2/src/main/org/h2/command/Parser.java

```diff
@@ -5198,73 +5204,122 @@ else if(readIf("CREATE")) {
                     WITH_STATEMENT_SUPPORTS_LIMITED_STATEMENTS);
         }
 
-        // clean up temp views starting with last to first (in case of
-        // dependencies)
-        Collections.reverse(viewsCreated);
-        p.setCteCleanups(viewsCreated);
+        // clean up temporary views starting with last to first (in case of
+        // dependencies) - but only if they are not persistent
+        if(!isPersistent){
+            p.setCteCleanups(viewsCreated);
+        }
         return p;
     }
 
-    private TableView parseSingleCommonTableExpression() {
-        String tempViewName = readIdentifierWithSchema();
+    @SuppressWarnings("resource")// Eclipse thinks targetSession needs releasing
+    private TableView parseSingleCommonTableExpression(boolean isPersistent) {
+        Session targetSession =  isPersistent ? database.getSystemSession() : session;
+        String cteViewName = readIdentifierWithSchema();
         Schema schema = getSchema();
-        Table recursiveTable;
+        Table recursiveTable=null;
         ArrayList<Column> columns = New.arrayList();
         String[] cols = null;
+        Database db = targetSession.getDatabase();
+        
+        System.out.println("systemSessionId="+database.getSystemSession().getId());
+        System.out.println("sessionId="+session.getId());
 
         // column names are now optional - they can be inferred from the named
-        // query if not supplied
+        // query, if not supplied by user
         if (readIf("(")) {
             cols = parseColumnList();
             for (String c : cols) {
-                // we don't really know the type of the column, so string will
+                // we don't really know the type of the column, so STRING will
                 // have to do
                 columns.add(new Column(c, Value.STRING));
             }
         }
-        Table old = session.findLocalTempTable(tempViewName);
-        if (old != null) {
-            if (!(old instanceof TableView)) {
+        
+        Table oldViewFound = null;
+        if(isPersistent){
+            oldViewFound = getSchema().findTableOrView(targetSession, cteViewName);
+        }
+        else{
+            oldViewFound = targetSession.findLocalTempTable(cteViewName);
+        }
+        if (oldViewFound != null && !isPersistent) {
+            if (!(oldViewFound instanceof TableView)) {
                 throw DbException.get(ErrorCode.TABLE_OR_VIEW_ALREADY_EXISTS_1,
-                        tempViewName);
+                        cteViewName);
             }
-            TableView tv = (TableView) old;
+            TableView tv = (TableView) oldViewFound;
             if (!tv.isTableExpression()) {
                 throw DbException.get(ErrorCode.TABLE_OR_VIEW_ALREADY_EXISTS_1,
-                        tempViewName);
+                        cteViewName);
+            }
+            if(isPersistent){
```

Comment:
this if block is not reachable because of checking !isPersistent earlier?

Reply:
You are right - removed outer condition. Thanks.